provisioning/containers/gui-common/websockify/instmetrics.go

```diff
@@ -0,0 +1,197 @@
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"log"
+	"math"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+// Resources utilization derived from ContainerMetrics.
+type Resources struct {
+	Connections []ConnInfo `json:"connections"`
+	CPUPerc     uint16     `json:"cpu"`
+	MEMPerc     uint16     `json:"mem"`
+	NET         int64      `json:"net"`
+	Timestamp   time.Time  `json:"timestamp,omitempty"`
+	ErrorMsg    string     `json:"error,omitempty"`
+}
+
+// InstanceMetricsHandler is the main handler for instance metrics.
+type InstanceMetricsHandler struct {
+	instanceMetricsClient *InstanceMetricsClient
+	// Application container resources.limits.cpu.
+	cpuLimit string
+	// Application container resources.limits.memory.
+	memoryLimit string
+	// PodName where application container is running.
+	podName string
+	// Last Resources extracted from CustomMetricsServer
+	cachedResources      *Resources
+	cachedResourcesMutex sync.RWMutex
+	// <UID, ConnInfo> map
+	connectionsTracking *sync.Map
+}
+
+// ServeHTTP serves WS server.
+func (h *InstanceMetricsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Println("upgrade error:", err)
+		return
+	}
+
+	var updatePeriod int64 = 2
+	if up, ok := r.URL.Query()["updatePeriod"]; ok {
+		updatePeriod, err = strconv.ParseInt(up[0], 10, 16)
+		if err != nil {
+			http.Error(w, "Invalid updatePeriod: it must be Integer (update seconds)", http.StatusBadRequest)
+			return
+		}
+	}
+	updatePeriodD := time.Duration(updatePeriod) * time.Second
+
+	var connUID *string
+	if uid, ok := r.URL.Query()["connUid"]; ok {
+		// noVNC page request
+		uids := uid[0]
+		connUID = &uids
+	} else {
+		// metricsDashboard request
+		connUID = nil
+	}
+
+	ip := r.Header.Get(headerXForwardedFor)
+	log.Printf("Incoming websocket connection on path /usages with update period %d from IP=%s", updatePeriod, ip)
+
+	go h.sendInstMetricsCycle(ws, updatePeriodD, connUID)
+}
+
+func (h *InstanceMetricsHandler) sendInstMetricsCycle(ws *websocket.Conn, updatePeriod time.Duration, connUID *string) {
+	ticker := time.NewTicker(updatePeriod)
+	for range ticker.C {
+		ctx, cancel := context.WithTimeout(context.Background(), updatePeriod)
+
+		if !h.cachedResourcesIsUpdated(updatePeriod) {
+			if err := h.updateCachedResources(ctx, updatePeriod); err != nil {
+				log.Printf("[sendCustomMetricsCycle] Error sending data on websocket. Closing WebSocket connection. Error: %v", err)
+				cancel()
+				ticker.Stop()
+			}
+		}
+		if err := h.sendCachedResources(ws, updatePeriod, connUID); err != nil {
+			log.Printf("[sendCustomMetricsCycle] Error sending data on websocket. Closing WebSocket connection. Error: %v", err)
+			cancel()
+			ticker.Stop()
+		}
+
+		// Operation running the context completed.
+		cancel()
```

Comment:
If you wrap the for body in a function you can defer the cancel() 
You could probably write the cycle inside the ServeHTTP method directly as an anonymous function, then leave this as the only send body